.github/actions/build-with-patched-std/action.yml

```diff
@@ -0,0 +1,48 @@
+# Github composite action to build a single-source-file test binary with an
+# already-checked-out version of Rust's stdlib, that will be patched with a
+# given revision of the backtrace crate.
+
+name: Build with patched std
+description: >
+  Build a binary with a version of std that's had a specific revision of
+  backtrace patched in.
+inputs:
+  backtrace-commit:
+    description: The git commit of backtrace to patch in to std
+    required: true
+  main-rs:
+    description: The (single) source code file to compile
+    required: true
+  rustc-dir:
+    description: The root directory of the rustc repo
+    required: true
+outputs:
+  test-binary-size:
+    description: The size in bytes of the built test binary
+    value: ${{ steps.measure.outputs.test-binary-size }}
+runs:
+  using: "composite"
+  steps:
+    - shell: bash
+      id: measure
+      env:
+        RUSTC_FLAGS: -Copt-level=s -Cstrip=symbols
```

Comment:
I do not want to use `-Copt-level=s` as the main check.
1. It's iffy! It sometimes results in larger sizes than 3, or even 2!
2. It is not representative of "default" `--release` builds, which are my primary concern.

Reply:
Example of how rustc sometimes is missing basic opts that would guarantee a smaller size for `-Copt-level=s`: https://github.com/rust-lang/rust/pull/100035

Reply:
I mean, one possibility is to do _both_ `-Os` and `-O` (default release), but then we're into matrix-style CI (I think?) and I'm not sure I'm up for doing that myself.

It's not a huge deal and I'm happy to change it back though. The code size varies between `-Os` and `-O3` _less_ than it does between architectures (IME). It's far more important to have good and well-defined sensitivity to things that affect code size, which as you say, `-Os` is probably not the best for.

Reply:
`-O` is **not** default `--release`, it is `-Copt-level=2`.

Reply:
`rustc --help` claims this currently: `    -O                  Equivalent to -C opt-level=2`.

Reply:
Oh, the default for `--release` is `-Copt-level=3`, I didn't realize. Maybe we should just create a new cargo project and build it with `--release` to have the most "realistic" situation?

Reply:
lmao am I the only one who knew?

yeah probably, if it's not too much additional effort.

Reply:
The case against that is that it's potentially a source of variability. Controlling the Rust flags directly rules out a sudden change in baseline because Cargo changed a default, or how it interprets a simplified option. (Granted, it would be a change in code size _between_ PRs, not in a single one, but it might still make accounting for changes difficult.)

Reply:
I think that this CI workflow is now specifically only designed to ever care about changes *within* a single PR. The size between PRs will change a lot anyways because of unrelated compiler/libstd/... changes.

Reply:
Strictly correct, but you might find it a little useful to be able to compare between PRs _somewhat_, even if it's not super scientific. Involving Cargo might trade that off a little.

If you _do_ want to go the route of using a full-fledged Cargo project, would it then be better to store it as actual files in the repo in some designated directory, instead of conjuring it out of a bash script embedded in a workflow YAML file like a devops matryoshka doll?

Reply:
I don't have a strong opinion beyond "it should roughly match the flags used by typical `--release` builds".

Reply:
@workingjubilee I am reverting the `opt-level` flag in my fork, but there is also the question of `strip`. Cargo's docs say [it defaults to `none`](https://doc.rust-lang.org/cargo/reference/profiles.html#strip), but I don't know if they mean "for the default profile of `debug`" or if that's true even for `release`. It's not listed in the [`release` profile settings](https://doc.rust-lang.org/cargo/reference/profiles.html#release).

Reply:
I would just do `cargo new --bin foo`, then overwrite `src/main.rs` and run `cargo +local build --release`. Before that we'd just need to `rustup toolchain link` the `build/<target>/stage0-sysroot` directory. At least it will also test that it can be linked with rustup :)

Reply:
It's not clear to me what the consensus is here (except to put `opt-level` back to 3). Would it be better to stick with the current single-file approach and implement the cargo-project-approach in a new PR?

Reply:
Let's switch to `-Copt-level=3` here and do the cargo changes in a new PR, it will require more changes (as explained in my previous comment).

Reply:
Just rebased with the `opt-level` changed back to 3 _but_ still with the additional strip option. Let me know if I should drop that one too.